security/authorization/engine/engine.go

```diff
@@ -0,0 +1,230 @@
+// i.e. if engine's action is ALLOW and match is true, we will return DecisionAllow;
+// if engine's action is ALLOW and match is false, we will return DecisionDeny.
+func getDecision(engine *policyEngine, match bool) Decision {
+	if engine.action == pb.RBAC_ALLOW && match || engine.action == pb.RBAC_DENY && !match {
+		return DecisionAllow
+	}
+	return DecisionDeny
+}
+
+// Returns the authorization decision of a single policy engine based on activation.
+// If any policy matches, the decision matches the engine's action, and the first
+//  matching policy name will be returned.
+// Else if any policy is missing attributes, the decision is unknown, and the list of
+//  policy names that can't be evaluated due to missing attributes will be returned.
+// Else, the decision is the opposite of the engine's action, i.e. an ALLOW engine
+//  will return DecisionDeny, and vice versa.
+func (engine *policyEngine) evaluate(activation interpreter.Activation) (Decision, []string) {
+	unknownPolicyNames := []string{}
+	for policyName, program := range engine.programs {
+		// Evaluate program against activation.
+		var match bool
+		out, _, err := program.Eval(activation)
+		if err != nil {
+			if out == nil {
+				// Unsuccessful evaluation, typically the result of a series of incompatible
+				// `EnvOption` or `ProgramOption` values used in the creation of the evaluation
+				// environment or executable program.
+				grpclog.Warning("Unsuccessful evaluation encountered during AuthorizationEngine.Evaluate: %s", err.Error())
+			}
+			// Unsuccessful evaluation or successful evaluation to an error result, i.e. missing attributes.
+			match = false
+		} else {
+			// Successful evaluation to a non-error result.
+			if !types.IsBool(out) {
+				grpclog.Warning("'Successful evaluation', but output isn't a boolean: %v", out)
+				match = false
+			} else {
+				match = out.Value().(bool)
+			}
+		}
+
+		// Process evaluation results.
+		if err != nil {
+			unknownPolicyNames = append(unknownPolicyNames, policyName)
+		} else if match {
+			return getDecision(engine, true), []string{policyName}
+		}
+	}
+	if len(unknownPolicyNames) > 0 {
+		return DecisionUnknown, unknownPolicyNames
+	}
+	return getDecision(engine, false), []string{}
+}
+
+// AuthorizationEngine is the struct for the CEL-based authorization engine.
+type AuthorizationEngine struct {
+	engines []*policyEngine
+}
+
+// NewAuthorizationEngine builds a CEL evaluation engine from a list of Envoy RBACs.
+func NewAuthorizationEngine(rbacs []*pb.RBAC) (*AuthorizationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return &AuthorizationEngine{}, status.Errorf(codes.InvalidArgument, "must provide 1 or 2 RBACs")
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return &AuthorizationEngine{}, status.Errorf(codes.InvalidArgument, "when providing 2 RBACs, must have 1 DENY and 1 ALLOW in that order")
+	}
```

Comment:
This seems hard to use.  Another idea:

```go
func NewAuthorizationEngine(allow, deny *pb.RBAC) (*AuthorizationEngine, error) {
	if allow == nil && deny == nil { error }
	if allow.Action != ALLOW || deny.Action != DENY { error }
	return &AuthorizationEngine{allow: newPolicyEngine(allow), deny: newPolicyEngine(deny)}
}
```

Not using a slice for the engines which have special requirements may also simplify the `Evaluate` function.